mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala

```diff
@@ -92,6 +92,56 @@ class RankingMetrics[T: ClassTag](predictionAndLabels: RDD[(Array[T], Array[T])]
     }.mean()
   }
 
+  /**
+   * Returns the mean average precision (MAP) at ranking position k of all the queries.
+   * If a query has an empty ground truth set, the average precision will be zero and a log
+   * warning is generated.
+   * @param k the position to compute the truncated precision, must be positive
+   * @return the mean average precision at first k ranking positions
+   */
+  @Since("3.0.0")
+  def meanAveragePrecisionAt(k: Int): Double = {
+    require(k > 0, "ranking position k should be positive")
+    predictionAndLabels
+      .map {
+        case (pred, lab) =>
+          averagePrecisionAt(pred, lab, k)
+      }
+      .mean()
+  }
+
+  /**
+   * Computes the average precision at first k ranking positions of all the queries.
+   * If a query has an empty ground truth set, the value will be zero and a log
+   * warning is generated.
+   *
+   * @param pred predicted ranking
+   * @param lab ground truth
+   * @param k use the top k predicted ranking, must be positive
+   * @return average precision at first k ranking positions
+   */
+  private def averagePrecisionAt(pred: Array[T], lab: Array[T], k: Int): Double = {
+    val labSet = lab.toSet
+
+    if (labSet.nonEmpty) {
+      var i = 0
+      var cnt = 0
+      var precSum = 0.0
+      val n = math.min(k, pred.length)
+      while (i < n) {
+        if (labSet.contains(pred(i))) {
+          cnt += 1
+          precSum += cnt.toDouble / (i + 1)
+        }
+        i += 1
+      }
+      precSum / math.min(labSet.size, k)
```

Comment:
I don't think the min is correct here? should divide through by the number of relevant docs in all cases.

Reply:
This is true when we are calculating `AP` but for `AP@k` I think we have to take `math.min(labSet.size, k)`. Cross checked with IR metrics literature and also found some reference related to AP@k in tensorflow https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/ops/metrics_impl.py#L2932
https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/ops/metrics_impl.py#L3052

Don't know whether my understanding is right? Please check once and let me know I will fix the above code.

Reply:
That's a good question. I do see that minimum, but it seems inconsistent with the wikipedia definition they point to (which is what I'm going on for Spark too). The 'number of relevant documents' doesn't depend on k. If it's capped at k then I think you'd end up with a weird looking precision-recall curve that would start at 1 when precision is 1, rather than 0. And average precision is the area under that curve.

Reply:
Let me come with some experiment to test and I will project that result. I think that will help.

Reply:
I ran the `RankingMetricsExample` with measured precisionAt & recallAt for k = [1, 2, 3, 4, 5] and also mapk with above implementation. I have plot precision-recall curves and also mapk curves. Need your inputs?

https://colab.research.google.com/drive/1plPwinJzce0tgK01T9ywtgV3JGW0dSYn#scrollTo=8Vdi2ISK8Nl7

Reply:
(I think precision is reversed, as it should go down and recall goes up, but I get it.)

I suppose that the "min(relevant, k)" definition only ends up being different by a constant factor. When computing all the relevant recall@i values, It's not defining recall@i as "correct / min(relevant, i)", which would be wrong -- would reduce to precision in most cases. But that's not what it's doing, OK.

The example at http://fastml.com/what-you-wanted-to-know-about-mean-average-precision/ however also supports dividing by "min(relevant, k)" so I think your approach is OK. Let me re-read the existing and new code then.